### PR TITLE
tests: enable pd test

### PR DESCRIPTION
**Pull request by overvenus**

### CODE REVIEW

File: tests/pd/test_rpc_client.rs [hunk 1]
```diff
@@ -24,12 +28,15 @@ use tikv::pd::{PdClient, RpcClient, validate_endpoints, Error as PdError, Region
 use super::mock::mocker::*;
 use super::mock::Server as MockServer;
 
+static PORT: AtomicUsize = AtomicUsize::new(52100);
```
Comment (siddontang) [inline tests/pd/test_rpc_client.rs]: can we use the way that listens port 0, gets a random port and then uses it?

I worry fixed PORT may be conflicted with other tests in raftstore which uses random port.

**Comment by siddontang** [comment]
Rest LGTM

File: tests/pd/test_rpc_client.rs (hunk 1 above)
Comment (overvenus) [inline tests/pd/test_rpc_client.rs]: grpc-rs need a port explicitly.
https://github.com/pingcap/grpc-rs/blob/eb657db414dcfe670bc949a965f15be31d6423b1/src/server.rs#L175

Comment (siddontang) [inline tests/pd/test_rpc_client.rs]: why must need the precise address?
Any better way to get around it?

**Comment by siddontang** [comment]
PTAL @BusyJay

**Comment by siddontang** [comment]
PTAL @disksing

File: tests/pd/mock/mocker/leader_change.rs [hunk 2]
```diff
@@ -32,13 +29,67 @@ struct Roulette {
...
+
+const DEAD_ID: u64 = 1000;
+const DEAD_NAME: &'static str = "walking_dead";
+const DEAD_URL: &'static str = "http://127.0.0.1:65534";
+
+impl Mocker for LeaderChange {
+    fn get_members(&self, _: &GetMembersRequest) -> Option<Result<GetMembersResponse>> {
+        let mut inner = self.inner.lock().unwrap();
+        let now = Instant::now();
+        if now.duration_since(inner.r.ts) > LeaderChange::get_leader_interval() {
```
Comment (BusyJay) [inline tests/pd/mock/mocker/leader_change.rs]: Why not `now - inner.r.ts`?

File: tests/pd/mock/server.rs [hunk 3]
```diff
@@ -12,54 +12,99 @@
...
-    _server: PDAsyncServer,
+    server: GrpcServer,
 }
 
 impl Server {
-    pub fn run<A, C>(addr: A, handler: Arc<Service>, case: Option<Arc<C>>) -> Server
+    pub fn run<C>(eps_count: usize, handler: Arc<Service>, case: Option<Arc<C>>) -> Server
+        where C: Mocker + Send + Sync + 'static
+    {
+        let eps: Vec<SocketAddr> = [0].iter()
```
Comment (BusyJay) [inline tests/pd/mock/server.rs]: You mean `vec![("127.0.0.1", 0); eps_count]`?

File: tests/pd/mock/server.rs [hunk 4]
```diff
@@ -12,54 +12,99 @@
...
     where R: Send + 'static,
           F: Fn(&Mocker) -> Option<Result<R>>
 {
-    if let Some(ref case) = mock.case {
-        match f(case.as_ref()) {
-            Some(Ok(resp)) => return futures::future::ok(resp).boxed(),
-            Some(Err(err)) => return futures::future::err(err).boxed(),
-            _ => (),
-        }
+    let mut resp = match mock.case {
```
Comment (BusyJay) [inline tests/pd/mock/server.rs]: How about `mock.case.as_ref().and_then(f).or_else(|| f(mock.handler.as_ref()))`?

File: tests/pd/mock/server.rs [hunk 5]
```diff
@@ -69,70 +114,108 @@ struct Mock<C: Mocker> {
     case: Option<Arc<C>>,
 }
 
-impl<C: Mocker> PDAsync for Mock<C> {
-    fn GetMembers(&self, req: GetMembersRequest) -> GrpcFutureSend<GetMembersResponse> {
-        try_takeover(self, |c| c.GetMembers(&req))
+impl<C: Mocker> Clone for Mock<C> {
+    fn clone(&self) -> Self {
+        let case = match self.case {
```
Comment (BusyJay) [inline tests/pd/mock/server.rs]: Why not `self.case.clone()`?

File: tests/pd/mock/server.rs [hunk 6]
```diff
@@ -33,12 +33,7 @@ impl Server {
     pub fn run<C>(eps_count: usize, handler: Arc<Service>, case: Option<Arc<C>>) -> Server
         where C: Mocker + Send + Sync + 'static
     {
-        let eps: Vec<SocketAddr> = [0].iter()
-            .cycle()
-            .take(eps_count)
-            .map(|_| ("127.0.0.1", 0).to_socket_addrs().unwrap().next().unwrap())
-            .collect();
-
+        let eps = vec![("127.0.0.1", 0).to_socket_addrs().unwrap().next().unwrap(); eps_count];
```
Comment (BusyJay) [inline tests/pd/mock/server.rs]: Why `to_socket_addrs`?

Comment (overvenus) [inline tests/pd/mock/server.rs]: https://doc.rust-lang.org/std/net/trait.ToSocketAddrs.html

`&'a [SocketAddrs]` impls `ToSocketAddrs`, not `Vec<(&str, u16)>` or `&'a [(&str, u16)]`.

**Comment by siddontang** [comment]
Any update? @overvenus

File: tests/pd/mock/mocker/bootstrap.rs [hunk 7]
```diff
@@ -18,16 +18,16 @@ use super::Result;
 use super::DEFAULT_CLUSTER_ID;
 
 #[derive(Debug)]
-pub struct AlreadyBootstrap {}
+pub struct AlreadyBootstrapped {}
```
Comment (BusyJay) [inline tests/pd/mock/mocker/bootstrap.rs]: `{}` is redundant.

File: tests/pd/mock/server.rs [hunk 8]
```diff
@@ -12,54 +12,89 @@
...
-        where A: ToSocketAddrs,
-              C: Mocker + Send + Sync + 'static
+    pub fn run<C>(eps_count: usize, handler: Arc<Service>, case: Option<Arc<C>>) -> Server
+        where C: Mocker + Send + Sync + 'static
+    {
+        let eps = vec![("127.0.0.1".to_owned(), 0); eps_count];
+        Server::run_with_eps(eps.as_slice(), handler, case)
+    }
+
+    pub fn run_with_eps<C>(eps: &[(String, u16)],
```
Comment (BusyJay) [inline tests/pd/mock/server.rs]: Why not just use `&[(&str, u16)]` instead?

File: tests/pd/mock/server.rs [hunk 9]
```diff
@@ -12,54 +12,89 @@
...
-            handler: handler,
-            case: case,
+            handler: handler.clone(),
+            case: case.clone(),
         };
-        Server { _server: PDAsyncServer::new(addr, Default::default(), m) }
+        let service = pdpb_grpc::create_pd(m);
+        let env = Arc::new(Environment::new(1));
+        let mut sb = ServerBuilder::new(env).register_service(service);
+        for ep in eps {
```
Comment (BusyJay) [inline tests/pd/mock/server.rs]: `for (host, port) in eps` seems more readable.

File: tests/pd/mock/mocker/bootstrap.rs [hunk 10]
```diff
@@ -18,11 +18,11 @@ use super::Result;
 use super::DEFAULT_CLUSTER_ID;
 
 #[derive(Debug)]
-pub struct AlreadyBootstrapped {}
+pub struct AlreadyBootstrapped;
 
 impl AlreadyBootstrapped {
     pub fn new() -> AlreadyBootstrapped {
-        AlreadyBootstrapped {}
+        AlreadyBootstrapped
```
Comment (BusyJay) [inline tests/pd/mock/mocker/bootstrap.rs]: `new` function is unnecessary.

**Comment by BusyJay** [comment]
LGTM

**Pull request merged**